### PR TITLE
Added DeepSHAP/DeepExplainer to FAQ

### PR DESCRIPTION
**Pull request by AvantiShri**
Includes detailed discussion of the pros/cons of the DeepLIFT-like implementations in DeepSHAP/DeepExplainer and DeepExplain (Ancona et al.)

### CODE REVIEW

**Comment by AvantiShri** [comment]
@slundberg you may be interested in looking this over - I've reproduced the relevant section below:

#### What are the similarities and differences between [DeepExplain from Ancona et al. (ICLR 2018)](https://github.com/marcoancona/DeepExplain) and [DeepSHAP/DeepExplainer from the SHAP repository](https://github.com/slundberg/shap)?
Both DeepExplain (Ancona et al.) and DeepSHAP/DeepExplainer work by overriding gradient operators, and can thus support a wider variety of architectures than those that are covered in the DeepLIFT repo (in fact, the DeepSHAP/DeepExplainer implementation was inspired by Ancona et al.'s work and builds on a connection between DeepLIFT and SHAP, described in the [SHAP paper](https://papers.nips.cc/paper/7062-a-unified-approach-to-interpreting-model-predictions)). For the set of architectures described in the DeepLIFT paper, i.e. linear matrix multiplications, convolutions, and single-input nonlinearities (like ReLUs), both these implementations are identical to DeepLIFT with the Rescale rule. However, neither implementation supports DeepLIFT with the RevealCancel rule (a rule that was developed to deal with failure cases such as the min function, and which unfortunately is not easily implemented by overriding gradient operators). The key differences are as follows:

(1) DeepExplain uses standard gradient backpropagation for elementwise operations (such as those present in LSTMs/GRUs/Attention). This will likely violate the summation-to-delta property (i.e. the property that the sum of the attributions over the input is equal to the difference-from-reference of the output). If you have elementwise operations, I recommend you use DeepSHAP/DeepExplainer, which employs a summation-to-delta-preserving backprop rule. The same is technically true for Maxpooling operations when a non-uniform reference is used (though this has not been a salient problem for us in practice); the DeepSHAP/DeepExplainer implementation guarantees summation-to-delta is satisfied for Maxpooling by assigning credit/blame to either the neuron that is the max in the actual input or the neuron that was the max in the reference (this is different from the 'Max' attribution rule proposed in the SHAP paper; that attribution rule does not scale well).

(2) DeepExplain (by Ancona et al.) does not support the dynamic reference that is demonstrated in the DeepLIFT repo (i.e. the case where a different reference is generated according to the properties of the input example, such as the 'dinucleotide shuffled' references used in genomics). [I've implemented the dynamic reference feature for DeepSHAP/DeepExplainer, with an associated example notebook](https://github.com/slundberg/shap/pull/322).

(3) DeepSHAP/DeepExplain is implemented such that multiple references can be used for a single example, and the final attributions are averaged over each reference. However, the way this is implemented, each GPU batch calculates attributions for [a single example, for all references](https://github.com/slundberg/shap/blob/98bf721cc063222feb2456504d20396eeba6a6aa/shap/explainers/deep/deep_tf.py#L257-L262). This means that the DeepSHAP/DeepExplainer implementation might be slow in cases where you have a large number of samples and only one reference. By contrast, DeepExplain (Ancona et al.) is structured such that the user provides a single reference, and this reference is used for all the examples. Thus, DeepExplain (Ancona et al.) allows GPU batching across examples, but does not allow for GPU batching across different references.

In summary, my recommendations are: use DeepSHAP if you have elementwise operations (e.g. GRUs/LSTMs/Attention), a need for dynamic references, or a large number of references compared to samples. Use DeepExplain when you have a large number of samples compared to references.

**Comment by slundberg** [comment]
@AvantiShri looks like a good summary to me. Thanks for sharing!
